Push notification listener

ggivo opened the pull request:
This PR enhances Jedis and UnifiedJedis to provide better support for RESP3 push messages and introduces a configurable option for relaxed timeout behavior.

**Background**

Starting with RESP3, Redis servers can send push messages to clients out-of-band (i.e., outside the standard request/response flow). See the [RESP3 Push Specification](https://redis.io/docs/latest/develop/reference/protocol-spec/#pushes) for more details.

In the current state of the client:
- 	Jedis supports only known **Push** notifications like **invalidate**.  Push notifications are supported using dedicated client-side caching (CSC) connection  (CacheConnection)
- 	This leads to runtime errors if `CLIENT TRACKING ON `is enabled in a context other than CSC, or if other **RESP3** push events are received.

✅ Changes Introduced
	•	Added support for **processing and discarding all RESP3 push messages**, preventing errors in non-CSC scenarios.
	•	Introduced support for **registering a custom PushListener**, enabling applications to handle push notifications as needed.
	•	**Pub/Sub push messages** are still handled as before and **propagated to the client** for backward compatibility.
	•	Added a **configurable option to enable relaxed timeout behavior**, intended for usage during maintenance events or when expecting slow responses from the server.

a-TODO-rov commented on src/main/java/redis/clients/jedis/JedisShardedPubSubBase.java:
Not much relevant changes in this file

a-TODO-rov commented on src/main/java/redis/clients/jedis/JedisClientConfig.java:
```diff
@@ -7,6 +7,8 @@
 
 import redis.clients.jedis.authentication.AuthXManager;
 
+import static redis.clients.jedis.PushHandler.NOOP;
```
Unused

a-TODO-rov commented on src/main/java/redis/clients/jedis/Protocol.java:
```diff
@@ -127,9 +129,9 @@ private static String[] parseTargetHostAndSlot(String clusterRedirectResponse) {
+  private static Object process(final RedisInputStream is, PushConsumer pushConsumer) {
     final byte b = is.readByte();
-    // System.out.println("BYTE: " + (char) b);
+    //System.out.println("BYTE: " + (char) b);
```
?

a-TODO-rov commented on src/main/java/redis/clients/jedis/util/JedisAsserts.java:
Not relevant changes here

ggivo commented on src/main/java/redis/clients/jedis/Connection.java:
```diff
@@ -614,4 +694,219 @@ protected boolean isTokenBasedAuthenticationEnabled() {
+  }
+
+
+  private static class MaintenanceEventConsumer implements PushConsumer {
```
I prefer to keep them private so we have more freedom in modifying the actual implementation in the future. 
Those consumers are not intended to be used by directly but via `MaintenanceEventListener` for example

ggivo commented on src/main/java/redis/clients/jedis/Protocol.java:
```diff
@@ -211,22 +214,52 @@ private static List<KeyValue> processMapKeyValueReply(final RedisInputStream is)
+    return reply;
   }
 
+  private static boolean isProcessed(PushConsumerContext reply) {
```
removed

ggivo commented on src/main/java/redis/clients/jedis/MaintenanceEventHandler.java:
```diff
@@ -0,0 +1,18 @@
+  void removeAllListeners();
+
+
+  Collection<MaintenanceEventListener> getListeners();
```
Introduced shared interface & abstract implementation

ggivo commented on src/main/java/redis/clients/jedis/PushHandler.java:
```diff
@@ -0,0 +1,71 @@
+ * @author Ivo Gaydajiev
+ * @since 6.1
+ */
+public interface PushHandler {
```
Introduced shared interface and abstract implementation with  MaintenanceEventHandler

ggivo commented on src/main/java/redis/clients/jedis/PushConsumerChain.java:
```diff
@@ -0,0 +1,210 @@
+
+  /**
+   * Handler that allows only pub/sub related events to be propagated to the client.
+   * Marks non-pub/sub events as processed, preventing their propagation.
```
Agree, not the best naming, I can update the comments, but opened for suggestions here.
Let's also discuss this.

My initial idea is that PushConsumers* are used only internally to process incoming PushNotifications so we can handle different types of existing Connections (CSC, Pub/Sub) in a backward compatible way. 
Then, to the end user of the library, we have `PushHandler` where we register `PushListeners` which propagate each `PushMessage` to the end user if interested.

ggivo commented:
We can simplify the logic by switching the default and removing the need for **CONSUME_ALL_HANDLER**.
**(Done)**

However, all applicable consumers must still be invoked if present. For example, if a user registers a listener for `maintenance events`, the `MaintenanceEventConsumer` must be called; likewise, we need the `ListenerNotificationConsumer` to notify any registered `PushListener`.

Until we refactor the Pub/Sub system and verify that no external code relies on propagated push events, we still need a mechanism to determine whether a given Push event should be propagated. Any consumer may update this decision, so the order of consumers remains important. 

Implementing an option for skipping a consumer is possible, but if propagation is disabled too early, again we end up with the same uncertainty. So I think the final resolution is to make sure no push messages are propagated at all, but don't think this can be handled easily and need to be tackled separately.

My idea is that **PushConsumers** should be kept internal and should not be exposed to users. Instead, users should interact with the system via PushListener.

ggivo commented:
> looks good!! there are couple of things not entirely clear to me. Just to clear them on my mind i will summarize as i understand; all push messages goes through all the consumers in the chain, in current impl, they go through;
> 
> * `CONSUME_ALL_HANDLER`
> * `PUBSUB_ONLY_HANDLER`
> * `MaintenanceEventConsumer`
> * `ListenerNotificationConsumer` - helps to plug customer app push listeners
> * `PushInvalidateConsumer` (in case of CSC -`CacheConnection`-)
> 
> what i don't like is each push messages visit all consumers no matter what. And the last one in the list, it kind of decides to set if data handed back to command execution or not. so we expect each consumer to do whatever it wants to do with message and set the context appropriately. this setup brings both flexibility and chaotic behaviour as well, because if someone shows up and marks it as they wish, they challenge to other consumer implementations plugged previously.
> 
> i think no push messages, other than pub/sub ones should be handed back to command executer anyway. (_and at some point, even they are not going to following the command response path in the case we introduce a way to customer app handle pub/sub messages via push listeners_). IMHO there should be something separate from letting client received/notified of push messages as well. May be with `PUBSUB_ONLY_HANDLER` functions the same way with `ListenerNotificationConsumer` but only for pub/sub messages.
> 
> let me try to explain the way i see it;
> 
> * a consumer chooses if each message propagates further to next consumer or not, and returns it as a value of consume result.
>   Each consumer along the chain should receive only what is let by its preceding consumers. this secures if messages should be seen or not, by any other consumer. Also this is more efficient in case there is a load of invalidation and/or pubsub messages.
> 
> ```java
> interface PushConsumer {
>    bool consume(PushMessageContext);
> }
> ```
> 
> * no messages other than pub/sub ones handed back to customer app (or received by command executor). isProcessed flag in context still remains the same to identify if message is chosen to be returned to the app
> * each push message initializes with default `isProcessed`=true and there is no need for a `CONSUME_ALL_HANDLER`. Then current naming of flag "isProcessed" might be suboptimal in this case. Alternatives would be `forwardToClient` or `isRequesterBound`

We can simplify the logic by switching the default and removing the need for **CONSUME_ALL_HANDLER**.
**(Done)**

However, all applicable consumers must still be invoked if present. For example, if a user registers a listener for `maintenance events`, the `MaintenanceEventConsumer` must be called; likewise, we need the `ListenerNotificationConsumer` to notify any registered `PushListener`.

Until we refactor the Pub/Sub system and verify that no external code relies on propagated push events, we still need a mechanism to determine whether a given Push event should be propagated. Any consumer may update this decision, so the order of consumers remains important. 

Implementing an option for skipping a consumer is possible, but if propagation is disabled too early, again we end up with the same uncertainty. So I think the final resolution is to make sure no push messages are propagated at all, but don't think this can be handled easily and need to be tackled separately.

My idea is that **PushConsumers** should be kept internal and should not be exposed to users. Instead, users should interact with the system via PushListener.

ggivo commented on src/main/java/redis/clients/jedis/Protocol.java:
```diff
@@ -211,22 +214,52 @@ private static List<KeyValue> processMapKeyValueReply(final RedisInputStream is)
     }
   }
 
   public static Object read(final RedisInputStream is) {
```
`read(is)` method is left for backward compatibility. In case anyone is using it outside the client code itself.
Will mark it as deprecated and remove in next major release.

ggivo commented:
Discussed with @atakavci  and we came up with the idea that it will be more optimal if instead of invoking the consumers, for each push message and each consumer decides if it is interested to process it, it will be better for consumers to register themself for a specific message type e.g. ("INVALIDATE") and get invoked only for it. 

Considering this an optimisation, we will try to improve it with a separate PR.

ggivo commented:
closed in favour of #4473

#4473 
- rebased on the latest code
- stripped down code not required by maintenance events 
   - removed support for user-provided listener for push notifications/maintenance events
